packages/caver-transaction/src/transactionTypes/abstractTransaction.js

```diff
@@ -0,0 +1,336 @@
+/*
+    Copyright 2020 The caver-js Authors
+    This file is part of the caver-js library.
+
+    The caver-js library is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    The caver-js library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with the caver-js. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+const _ = require('lodash')
+const Bytes = require('eth-lib/lib/bytes')
+const RLP = require('eth-lib/lib/rlp')
+const Hash = require('eth-lib/lib/hash')
+const TransactionHasher = require('../transactionHasher/transactionHasher')
+const utils = require('../../../caver-utils')
+const Keyring = require('../../../caver-wallet/src/keyring/keyring')
+const {
+    TX_TYPE_STRING,
+    TX_TYPE_TAG,
+    refineSignatures,
+    typeDetectionFromRLPEncoding,
+    isValidNumber,
+} = require('../transactionHelper/transactionHelper')
+const { KEY_ROLE } = require('../../../caver-wallet/src/keyring/keyringHelper')
+const { validateParams } = require('../../../caver-core-helpers/src/validateFunction')
+
+/**
+ * Abstract class that implements common logic for each transaction type.
+ * @class
+ */
+class AbstractTransaction {
+    /**
+     * Abstract class that implements common logic for each transaction type.
+     * In this constructor, type, tag, nonce, gasPrice, chainId, gas and signatures are set as transaction member variables.
+     *
+     * @constructor
+     * @param {string} typeString - The type string of transaction.
+     * @param {string} typeTag - The type tag in format of hex string.
+     * @param {object} createTxObj - The parameters to create an instance of transaction.
+     */
+    constructor(typeString, typeTag, createTxObj) {
```

Comment:
Do we need to pass both typeString and typeTag? I think one can be derived by the other.

Reply:
I removed typeTag parameter. :)